route/auth-router.js

```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+const jsonParser = require('body-parser').json();
+const debug = require('debug')('cfgram:auth-router');
+const Router = require('express').Router;
+const basicAuth = require('../lib/basic-auth-middleware.js');
+
+const User = require('../model/user.js');
+
+const authRouter = module.exports = Router();
+
+authRouter.post('/api/signup', jsonParser, function(req, res, next) {
+  debug('POST /api/signup');
+
+  let password = req.body.password;
+  delete req.body.password;
+
+  let user = new User(req.body);
+
+  user.generatePasswordHash(password)
+  .then( user => user.save())
+  .then( user => user.generateToken())
+  .then( token => res.send(token))
+  .catch(next);
+});
+
+authRouter.get('/api/signin', basicAuth, function(req, res, next) {
+  debug('GET /api/signin');
+
+  User.findOne({ username: req.auth.username })
+  .then( user => user.comparePasswordHash(req.auth.password))
```

Comment:
if mongoose can't find a user based on the username you passed it then the `findOne()` function will return null and move on to the promise which will then error out in the `comparePasswordHash` because the username is null. You need to create error middleware to handle this.